src/main/java/com/atlassian/db/replica/api/ThrottledCache.java

```diff
@@ -57,36 +62,106 @@ private ThrottledCacheWithTimeout(Clock clock, Duration timeout) {
         @Override
         public Optional<T> get(Supplier<T> supplier) {
             maybeRefresh(supplier);
-            return Optional.ofNullable(value);
+            return Optional.ofNullable(lockValuePair.get().getValue());
         }
 
         @Override
         public Optional<T> get() {
-            return Optional.ofNullable(value);
+            return Optional.ofNullable(lockValuePair.get().getValue());
         }
 
         private void maybeRefresh(Supplier<T> supplier) {
             final Instant deadline = clock.instant().plus(this.timeout);
-            final boolean hasLock = lock.compareAndSet(null, deadline);
+            final T currentValue = this.lockValuePair.get().getValue();
+            final LockValuePair<T> valueWithLock = new LockValuePair<>(
+                deadline,
+                currentValue
+            );
+            final LockValuePair<T> currentLockValuePair = this.lockValuePair.get();
+            final boolean hasLock = !currentLockValuePair.isLocked() && this.lockValuePair.compareAndSet(
+                currentLockValuePair,
+                valueWithLock
+            );
             if (hasLock) {
-                loadValue(supplier, deadline);
+                loadValue(supplier, valueWithLock);
             } else {
-                final Instant instant = lock.get();
-                final boolean didLockExpire = instant != null && instant.isBefore(clock.instant());
-                if (didLockExpire) {
-                    final boolean didLockTakeOver = lock.compareAndSet(instant, deadline);
+                final LockValuePair<T> oldLock = this.lockValuePair.get();
+                if (oldLock.didLockExpire(clock)) {
+                    final LockValuePair<T> newLock = new LockValuePair<>(
+                        deadline,
+                        oldLock.getValue()
+                    );
+                    final boolean didLockTakeOver = lockValuePair
+                        .compareAndSet(oldLock, newLock);
                     if (didLockTakeOver) {
-                        loadValue(supplier, deadline);
+                        loadValue(supplier, newLock);
```

Comment:
There's one more thing that is tempting to implement. To `interrupt` the slow thread here (we are going to ignore the results anyway). The only thing that stops me is that I know it will be a database call and database calls are not interruptable, so there may be no benefit from the extra complexity.

Reply:
> are not interruptable


well you never know, even if this / or sth else happens you can improve it along the way

Reply:
DB calls can have timeout ;)